include/rocksdb/multi_scan_iterator.h

```
@@ -0,0 +1,247 @@
+    if (!iter_->Valid()) {
+      empty_ = true;
+    }
+    status_ = iter_->status();
+  }
+
+  void SeekNext() {
+    idx_++;
+    if (idx_ < scans_.size()) {
+      NextScan();
+    } else {
+      empty_ = true;
+    }
+  }
+
+  bool empty() { return empty_; }
+
+  Status status() { return status_; }
+
+ private:
+  const std::vector<ScanDesc>& scans_;
+  size_t idx_;
+  std::unique_ptr<Iterator> iter_;
+  bool empty_;
+  Status status_;
+
+  void NextScan() {
+    iter_->Seek(scans_[idx_].start);
+    empty_ = !iter_->Valid();
+    status_ = iter_->status();
+  }
+};
+#endif
+
+// An iterator that returns results from multiple scan ranges. The ranges are
+// expected to be in increasing sorted order. The application on top of RocksDB
+// would use this as follows -
+//
+//  std::vector<ScanDesc> scans{{.start = Slice("bar")},
+//                              {.start = Slice("foo")}};
+//  std::unique_ptr<MultiScanIterator> iter.reset(
+//                                      db->NewMultiScanIterator());
+//  try {
+//    for (auto scan = iter.begin(); scan != iter.end(); ++scan) {
+//      for (auto kv = scan.begin(); kv != scan.end(); ++kv) {
+//        auto kvpair = *kv;
+//        // Do something with key - kv.first
+//        // Do something with value - kv.second
+//      }
+//    }
+//  } catch {
+//  }
+class MultiScanIterator {
+ public:
+  MultiScanIterator(const std::vector<ScanDesc>& scans,
+                    std::unique_ptr<Iterator>&& db_iter)
+      : scans_(scans), db_iter_(std::move(db_iter)) {}
+
+  class ScanIterator {
+   public:
+    class Scan;
+
+    using self_type = ScanIterator;
+    using value_type = Scan;
+    using reference = Scan&;
+    using pointer = Scan*;
+    using difference_type = int;
+    using iterator_category = std::input_iterator_tag;
+
+    ScanIterator(const std::vector<ScanDesc>& scans, Iterator* db_iter)
+        : scans_(scans), idx_(0), db_iter_(db_iter) {
+      db_iter_->Seek(*scans_[idx_++].range.start);
```

do we want to throw InvalidArg if `idx_ >= scans_.size()`?